src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs

```diff
@@ -183,19 +226,35 @@ public ValueTask<FlushResult> WriteChunkAsync(ReadOnlySpan<byte> buffer, Cancell
 
                 // Make sure any memory used with GetMemory/Advance is written before the chunk
                 // passed in.
-                WriteCurrentMemoryToPipeWriter();
-
-                if (buffer.Length > 0)
+                if (_advancedBytesForChunk > 0 || buffer.Length > 0)
                 {
                     var writer = new BufferWriter<PipeWriter>(_pipeWriter);
+                    if (_advancedBytesForChunk > 0)
+                    {
+                        WriteCurrentMemoryToPipeWriter(ref writer);
+                    }
+                    else
+                    {
+                        // If there is an empty write, we still need to update the current chunk
+                        _currentChunkMemoryUpdated = false;
+                    }
 
-                    writer.WriteBeginChunkBytes(buffer.Length);
-                    writer.Write(buffer);
-                    writer.WriteEndChunkBytes();
-                    writer.Commit();
+                    if (buffer.Length > 0)
+                    {
 
+                        writer.WriteBeginChunkBytes(buffer.Length);
+                        writer.Write(buffer);
+                        writer.WriteEndChunkBytes();
+                    }
+
+                    writer.Commit();
                     _unflushedBytes += writer.BytesCommitted;
                 }
+                else
+                {
+                    // If there is an empty write, we still need to update the current chunk
+                    _currentChunkMemoryUpdated = false;
```

Comment:
_currentChunkMemoryUpdated is always false by the end of this method. I'd do the following:
- remove setting _currentChunkMemoryUpdated = false in WriteCurrentMemoryToPipeWriter()
- set _currentChunkMemoryUpdated  to false in WriteChunkAsync and FlushAsync

Reply:
WriteChunkAsync calls FlushAsync, so only need to do it there?

Reply:
Yeah I think you are right.

Reply:
Done